[Gekidou - MM-44928] Search Screen - Search Modifiers

jfrerich opened the pull request:
#### Summary
This PR adds the Search Modifiers to the search Screen. 

Corner Cases: 
*  clear search
   * user has performed a search and hits the `x` in the navigation search header
   * text is cleared and modifiers is shown
* unlimited amount of modifiers allowed
   * user can add `from:A from:B in:xyz in:abc` 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-44928

#### Checklist
<!--
Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.
-->
- [x] Has UI changes
- [x] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: <!-- Device name(s), OS version(s) -->

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs/Videos (for both iOS and Android if possible).
-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* New features and improvements, including behavioural changes, UI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Example:

```release-note
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->

```release-note
NONE
```

jfrerich commented:
@matthewbirtch, for this PR, I am mostly concerned about the corner cases listed in the PR description. This PR performs like V1 with regards to adding modifiers. You can add as many as you'd like by tapping a modifier many times. 

Another concern is what to do with the back button in the search navigation header. My gut instinct is that tapping the back button should show the previous search results?

matthewbirtch commented on app/screens/home/search/modifiers/show_more.tsx:
```diff
@@ -0,0 +1,49 @@
+        showMore: {
+            color: theme.buttonBg,
+            paddingLeft: 20,
+            paddingVertical: 12,
```
I think you can remove the vertical padding. It seems to inherit the correct padding already without this

matthewbirtch reviewed (changes requested):
@jfrerich I added some comments on changes to the typography and spacing.

As for the corner cases...

> clear search
user has performed a search and hits the x in the navigation search header
text is cleared and modifiers is shown

This corner case is behaving correctly. Nice!

> unlimited amount of modifiers allowed
user can add from:A from:B in:xyz in:ABC

This is fine for now....I don't see too many cases where users would want to user more than 3 modifiers though. We could consider limiting to 3 modifiers, but we don't do that in v1, so we can leave it as-is for now

jfrerich commented on app/screens/home/search/modifiers/modifiers.tsx:
```diff
@@ -0,0 +1,187 @@
+            paddingVertical: 12,
+            paddingHorizontal: 20,
+            color: theme.centerChannelColor,
+            ...typography('Heading', 600, 'SemiBold'),
```
Figma has this at 300, not 200.

matthewbirtch replied:
That's my mistake, you're right about 300

jfrerich commented:
> Are you still going to be hooking up the autocompletes when a modifier is used? Will that be another PR?
Yes, I will hook up autocomplete and it will be in a separate PR.

matthewbirtch reviewed (approved):
modifiers look good now

larkox commented on app/screens/home/search/modifiers/show_more.tsx:
```diff
@@ -0,0 +1,48 @@
+            onPress={onPress}
+            labelComponent={
+                <FormattedText
+                    id={showMore ? 'mobile.search.show_less' : 'mobile.search.show_more'}
+                    defaultMessage={showMore ? 'Show less' : 'Show more'}
```
This probably won't get found by i18n-extract. Either use the t function or refactor it so it is something like:
```
let labelComponent;
if (showMore) {
    labelComponent = <FormattedText id={'mobile.search.show_less'} defaultMessage={...} ... />
} else {
    labelComponent = <Formatted Text ... />
}
```

larkox reviewed (changes requested):
Great work. Just some minor changes, and you also need to run i18n-extract.

jfrerich commented:
@matthewbirtch, FYI, I added an `After:` modifier which is not shown in figma.

jfrerich commented on app/screens/home/search/modifiers/modifier.tsx:
```diff
@@ -0,0 +1,106 @@
+                </View>
+            }
+
+            i18nId={item.id}
```
The id is actually not needed in this case because all of the text in the menu item is added via the `labelComponent` prop. And that was needed because of the bold and non-bolded text.

larkox commented:
@jfrerich run the `npm run i18n-extract` so it gets the new strings in the relevant file.

jfrerich commented:
@larkox, extract is failing right now. 
<img width="1030" alt="image" src="https://user-images.githubusercontent.com/7575921/174347109-97b36011-721a-4c54-b72e-9af70a1e7ad8.png">

larkox commented:
@jfrerich `mkdir share_extension` should do the trick ;)

jfrerich commented:
@larkox, I made some other edits to the i18n names but shorting them and using `modifier` as a prefix so they get grouped. I also removed the leading space and added it in the component instead.

larkox commented on assets/base/i18n/en.json:
```diff
@@ -508,6 +508,15 @@
+  "mobile.search.modifier_in": "a specific channel",
+  "mobile.search.modifier_on": "a specific date",
+  "mobile.search.modifier_phrases": "messages with phrases",
+  "mobile.search.show_less": "",
+  "mobile.search.show_more": "",
```
These don't have a text. Probably because they use the `t` function, in which case you have to manually set the text here.

larkox commented on app/screens/home/search/modifiers/modifiers.tsx:
```diff
@@ -65,31 +65,31 @@ const getModifiersSectionsData = (intl: IntlShape): ModifierItem[] => {
             term: 'From:',
             testID: 'search.from_section',
-            description: formatMessage({id: 'mobile.search.from_modifier_description', defaultMessage: ' a specific user'}),
+            description: formatMessage({id: 'mobile.search.modifier_from', defaultMessage: ' a specific user'}),
```
0/5 I would use a `.` instead of a `_`, but it is just personal preference.

larkox reviewed (changes requested):
Blocking just because of the missing texts. Rest looks fine.

enahum commented on app/screens/home/search/modifiers/modifiers.tsx:
```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
you should rename this file from `modifiers.tsx` to `index.tsx`

enahum commented on app/screens/home/search/modifiers/show_more.tsx:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    return (
+        <MenuItem
```
I believe `MenuItem` is being replace with `OptionItem` not 100% sure if that applies here.

@avinashlng1080 thoughts? @jfrerich can you try replacing this with `OptionItem`

enahum commented on app/screens/home/search/modifiers/modifier.tsx:
```diff
@@ -0,0 +1,101 @@
+    });
+
+    return (
+        <MenuItem
```
see my comment about `OptionItem`

jfrerich replied:
Heres the output using the `OptionItem` component.  I added an `inline` boolean prop which places the description inline with the label.  @matthewbirtch, the styling is slightly different than the figma design. @avinashlng1080, are you ok with my changes to `OptionItem` and the new `inline` prop?

@enahum, thanks for the heads up. This reduces the `Modifier` component code significantly.

Figma Design: 
<img width="402" alt="image" src="https://user-images.githubusercontent.com/7575921/176247763-b3d2e05b-3f65-49ea-9adc-04fbdb03d0a9.png">

Using OptionItem component:
<img width="514" alt="image" src="https://user-images.githubusercontent.com/7575921/176247600-b7c0940a-be19-485e-b484-8d47e6559b54.png">

enahum replied:
I guess we could pass additional styles to the `OptionItem` component to make it match the design or we can keep the keep MenuItem and then refactor once MenuItem is removed from the codebase

jfrerich replied:
I think passing styles to the `OptionItem` would be nice, but also allow alot of variabilty in it's look. 

I could also adjust the logic when `inline` is set:

* if inline === false:
  * use the default optionItem ServiceSettings
* if inline === true and description = '':
  * use default optionItem ServiceSettings
* if inline === true && description:
  * use output style in Figma.

jfrerich replied:
I can revert the change to use OptionItem, but I am 1/5 we keep it

matthewbirtch replied:
If we can keep the styles, that's ideal, but if it adds too much bloat to the code then I would be ok with the way it looks in your screenshot above @jfrerich

jfrerich replied:
I added some logic to the component that will render a one-line option as shown in the figma design. If the `inline` prop is true and `description` is not empty, the Option item text will render the same as desired in the Figma design.

enahum commented on app/components/option_item/index.tsx:
```diff
@@ -79,12 +93,34 @@ const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
     const theme = useTheme();
     const styles = getStyleSheet(theme);
 
+    const useInLine = useMemo(() => {
```
The name of the var can create confusion with a hook. But do we need to memoize it? Maybe call it isInline?

larkox replied:
No need to memoize at all. It is a base type (Boolean) that is probably practically the same amount of process than the overhead of useMemo (and less memory usage).

enahum reviewed (approved):
Approved but if you can the change on the things I've comment that'll be great

larkox commented on app/components/option_item/index.tsx:
```diff
@@ -79,12 +93,34 @@ const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
+
+    const labelStyle = useMemo(() => {
+        return useInLine ? styles.inlineLabel : styles.label;
+    }, [inline, styles, useInLine]);
```
inline is not used in the body of the memo, so I do not see any reason to have it in the dependencies. The same for the next two functions.

jfrerich replied:
Removed in 5472b2579f694ccb06f222f1ebda965a074feb1f

larkox commented on app/screens/home/search/modifiers/index.tsx:
```diff
@@ -0,0 +1,157 @@
+
+    const handleShowMore = useCallback(() => {
+        setShowMore(!showMore);
+        show.value = (showMore ? 3 : data.length) * MODIFIER_LABEL_HEIGHT;
```
(0/5) This is a bit confusing. For readability sake I would go with something like:
```suggestion
        const nextShowMore = !showMore;
        show.value = (nextShowMore ? data.length : 3) * MODIFIER_LABEL_HEIGHT;
        setShowMore(nextShowMore);
```

larkox commented on app/screens/home/search/search.tsx:
```diff
@@ -130,11 +133,12 @@ const SearchScreen = ({teamId}: Props) => {
         setFilter(FileFilters.ALL);
-    }, [filter]);
+        setShowResults(false);
+    }, [filter, searchValue, lastSearchedValue]);
```
Not sure why this used to have the `filter` as a dependency, but seems like it depends on nothing.

jfrerich replied:
@larkox, shouldn't it be dependent on `[setSearchValue, setLastSearchedValue, setFilter, setShowResults]`?

larkox reviewed (approved):
Some minor changes and should be ready to go.
